### PR TITLE
#167212749 create page layout

### PR DESCRIPTION
**Pull request by mosinmiloluwa01**
#### What does this PR do?
This PR creates page layout

#### Description of task to be completed?
- create page layout template
- create header content molecule
- create header organism
- add header to page layout

#### How should this be manually tested?
- run `npm install`
- run `npm start` on your console

#### Any background context you want to provide?
user must have node installed

#### What are the relevant pivotal tracker stories?
[167212749](https://www.pivotaltracker.com/story/show/167212749)

#### Screenshots (if appropriate)
Screenshot 2019-07-16 at 10 34 40 AM

### CODE REVIEW

**Review by akinyeleolat — APPROVED** [review]

**Comment by akinyeleolat** [comment]
Fix the merge conflicts. It Looks Good to me.

**Review by Sojisoyoye — APPROVED** [review]

File: src/components/UI/atoms/InputField/InputField.jsx [hunk 1]
```diff
@@ -73,6 +72,7 @@ InputField.propTypes = {
   fontSize: PropTypes.oneOf(Object.keys(fontSizes)),
   inputWidth: PropTypes.oneOf(Object.keys(width)),
   borderRadius: PropTypes.oneOf(Object.keys(themeBorderRadius)),
+  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
```
Comment (benfluleck) [inline src/components/UI/atoms/InputField/InputField.jsx]: Why would this ever be a number 🤔

File: src/components/UI/molecules/Headers/HeaderNavigation.jsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import React, { Fragment } from 'react';
+import PropTypes from 'prop-types';
+import FlexContainer from '<atoms>/layouts/FlexContainer/FlexContainer';
+
+const HeaderNavigation = ({ navItems }) => {
```
Comment (benfluleck) [inline src/components/UI/molecules/Headers/HeaderNavigation.jsx]:
```suggestion
const HeaderNavigation = ({ navItems }) => (
    <FlexContainer
      padding="zero"
      justifyContent="flex-end"
      containerWidth="searchBarWidth"
      alignItems="flex-end"
      flexDirection="row"
    >
      <Fragment>{navItems.map(items => items)}</Fragment>
    </FlexContainer>
  );
```

File: src/components/UI/atoms/Button/Button.jsx [hunk 3]
```diff
@@ -17,21 +18,31 @@ const Button = ({
...
     display={display}
     onClick={onClick}
+    marginLeft={marginLeft}
+    padding={padding}
   >
     {children}
   </Button.Container>
 );
 
+const { spacing } = themes;
```
Comment (benfluleck) [inline src/components/UI/atoms/Button/Button.jsx]: Put this at line 5

File: src/components/UI/molecules/Headers/HeaderNavigation.spec.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import '@testing-library/jest-dom/extend-expect';
+
+afterEach(cleanup);
+
+const navItems = headerMapper(items);
+
+describe('Header Navigation', () => {
+  it('should render header navigation component', async () => {
+    const { container } = render(
+      <MemoryRouter>
```
Comment (benfluleck) [inline src/components/UI/molecules/Headers/HeaderNavigation.spec.js]: This should be abstracted out to a setup function

File: src/components/UI/organisms/Header/Header.jsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import HeaderNavigation from '<molecules>/Headers/HeaderNavigation';
+import Title from '<atoms>/Title/Title';
+import FlexContainer from '<atoms>/layouts/FlexContainer/FlexContainer';
+
+const Header = ({ navItems }) => {
```
Comment (benfluleck) [inline src/components/UI/organisms/Header/Header.jsx]: Please remove the return as shown above

File: src/components/UI/organisms/Header/Header.spec.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+afterEach(cleanup);
+
+const navItems = headerMapper(items);
+
+describe('Header', () => {
+  it('should render Header component', () => {
+    const { container } = render(
+      <MemoryRouter>
+        <Header navItems={navItems}>
+          <Text>Hello world</Text>
```
Comment (benfluleck) [inline src/components/UI/organisms/Header/Header.spec.js]: This should be in a setup function

File: src/components/pages/homepage/Homepage.jsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import PageLayout from '../../templates/pageLayout';
+import Text from '<atoms>/Text/Text';
+import headerMapper from '<helpers>/headerMapper';
+import loadData from '<helpers>/dummyLoader';
+import items from './navItems';
+
+const Homepage = () => {
+  const navItems = headerMapper(items);
+
+
```
Comment (benfluleck) [inline src/components/pages/homepage/Homepage.jsx]: Remove the space please

File: src/components/pages/homepage/Homepage.jsx [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import PageLayout from '../../templates/pageLayout';
```
Comment (benfluleck) [inline src/components/pages/homepage/Homepage.jsx]: Keep the casing consistent for the folder structure /HomePage/HomePage.jsx

File: src/components/pages/homepage/Homepage.jsx [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        textAlign="center"
+        color="primary"
+      >
+        Welcome to Authors Haven
+      </Text>
+    </PageLayout>
+  );
+};
+
+export default Homepage;
```
Comment (benfluleck) [inline src/components/pages/homepage/Homepage.jsx]:
```suggestion
export default HomePage;
```

File: src/components/templates/PageLayout.spec.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+afterEach(cleanup);
+
+const navItems = headerMapper(items);
+
+describe('Page Layout', () => {
+  it('should render page layout with initial state and callback', () => {
+    const { container, getByPlaceholderText, getByText } = render(
+      <MemoryRouter>
+        <PageLayout navItems={navItems} initialState={true} callback={loadData}>
+          <Text
```
Comment (benfluleck) [inline src/components/templates/PageLayout.spec.js]: Text with props not with html elements

File: src/components/templates/pageLayout.jsx [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        <Header navItems={navItems} />
+        <PageLayout.Container>{children}</PageLayout.Container>
+      </Fragment>
+    );
+  }
+
+  return (
+    dataLoader() && (
+      <PageLayout.Container justifyContent="center" alignItems="center">
+        <Loader type="Rings" color="#B02091" height="100" width="100" />
```
Comment (benfluleck) [inline src/components/templates/pageLayout.jsx]: Abstract this in one component and import it as Loader

File: src/components/templates/pageLayout.jsx [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+ * @prop {children} children - children
+ * @prop {navItems} children - navigation items
+ *
+ * @return {Header} Page layout
+ */
+
+const PageLayout = ({
+  children, navItems, initialState, callback,
+}) => {
+  const [isLoaded, setIsLoaded] = useState(initialState);
```
Comment (benfluleck) [inline src/components/templates/pageLayout.jsx]: A better name would be `isLoading` and `setIsLoading`

File: src/components/templates/pageLayout.jsx [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+  const dataLoader = async () => {
```
Comment (benfluleck) [inline src/components/templates/pageLayout.jsx]: Please rename this like a function

File: src/components/templates/pageLayout.jsx [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    display: flex;
+    justify-content: ${justifyContent};
+    align-items: ${alignItems};
+`}
+`;
+
+PageLayout.defaultProps = {
+  initialState: 'true',
+  display: 'flex',
+  callback: () => {
```
Comment (benfluleck) [inline src/components/templates/pageLayout.jsx]:
```suggestion
  callback: () => {}
```

File: src/core/routes.js [hunk 15]
```diff
@@ -1,4 +1,4 @@
-import HomePage from '<pages>/Homepage';
+import HomePage from '<pages>/homepage/Homepage';
```
Comment (benfluleck) [inline src/core/routes.js]: Please keep the naming connsistent

File: src/helpers/dummyLoader.js [hunk 16]
```diff
@@ -0,0 +1,7 @@
+const loadData = async (millisecond = 3000) => {
```
Comment (benfluleck) [inline src/helpers/dummyLoader.js]: What is this doing here

File: src/helpers/headerMapper.js [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+import Text from '<atoms>/Text/Text';
+import InputField from '<atoms>/InputField/InputField';
+import StyledLink from '<atoms>/StyledLink/StyledLink';
+import Button from '<atoms>/Button/Button';
+
+const headerMapper = links => Object.keys(links).map((component, i) => {
+  switch (links[component].type) {
+  case 'link':
+    return (
+      <StyledLink key={i} marginleft="sm" to={links[component].action}>
```
Comment (benfluleck) [inline src/helpers/headerMapper.js]: index not i

File: src/helpers/headerMapper.js [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+        padding="xs"
+        onChange={links[component].action}
+        onFocus={(e) => {
+          e.target.placeholder = '';
+        }}
+      />
+    );
+  case 'button':
+    return (
+      <StyledLink key={i} marginleft="sm" to={links[component].location}>
```
Comment (benfluleck) [inline src/helpers/headerMapper.js]: index not i

File: src/helpers/headerMapper.js [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+        <Button
+          as={Link}
+          key={i}
```
Comment (benfluleck) [inline src/helpers/headerMapper.js]: why does the button take a key

File: src/helpers/headerMapper.js [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+          textTransform="capitalize"
+        >
+          {component}
+        </Text>
+      </StyledLink>
+    );
+  case 'input':
+    return (
+      <InputField
+        key={i}
```
Comment (benfluleck) [inline src/helpers/headerMapper.js]: Why is the key and id the same

**Review by benfluleck — CHANGES_REQUESTED** [review]

File: src/components/templates/PageLayout/PageLayout.jsx [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * @prop {children} children - children
+ * @prop {navItems} children - navigation items
+ *
+ * @return {Header} Page layout
+ */
+const PageLayout = ({ children, navItems, callback }) => {
+  const [isLoading, setIsLoaded] = useState(true);
+
+  const retrieveData = async () => {
```
Comment (benfluleck) [inline src/components/templates/PageLayout/PageLayout.jsx]: This is too generic, retrieve what Data ?

Comment (benfluleck) [reply src/components/templates/PageLayout/PageLayout.jsx]: pageData

**Review by benfluleck — APPROVED** [review]

**Pull request merged**